doc/cephfs/lazyio.rst

```diff
@@ -21,3 +21,43 @@ LazyIO can be enabled by following ways.
 
 - ``ioctl(fd, CEPH_IOC_LAZYIO, 1UL)`` enables LAZY_IO for file handle in
    ceph-fuse mount.
+
+Using LazyIO
+============
+
+ LazyIO includes two methods ``lazyio_propagate()`` and ``lazyio_synchronize()``. With LazyIO enabled, writes may not be visble to other clients until ``lazyio_propagate()`` is called. Reads may come from local cache (irrespective of changes to the file by other clients) until ``lazyio_synchronize()`` is called.
+
+- ``lazyio_propagate(...)`` - Ensures that any buffered writes of the calling client, in the specific region, has been propagated to the shared file.
+
+- ``lazyio_synchronize(...)`` - Ensures that the calling client is, in a subsequent read call, able to read the updated file with all the propagated writes of the other clients. In CephFS this is facilitated by invalidating the inode of the file and hence forces the client to refetch/recache the data from the updated file. Also if the write cache of the calling client is dirty(not propagated), lazyio_synchronize() flushes it as well.
+
+An example usage(utilizing libcephfs) is given below. Here the two client mounts are ca and cb.
+
+::
+
+        /* ca and cb open the shared file file.txt */
+        int fda = ceph_open(ca, "file.txt", O_CREAT|O_RDWR, 0644); 
+        int fdb = ceph_open(ca, "file.txt", O_CREAT|O_RDWR, 0644);
+
+        /* Enable LazyIO on both clients */
+        ceph_lazyio(ca, fda, 1));
+        ceph_lazyio(cb, fdb, 1));
+
+        char out_buf[] = "fooooooooo";
+        
+        /* ca and cb issue a write and propagates it respectively */
+        ceph_write(ca, fda, out_buf, sizeof(out_buf), 0);*/
+        ceph_propagate(ca, fda, 0, 0);
+
+        ceph_write(cb, fdb, out_buf, sizeof(out_buf), 10));
+        ceph_lazyio_propagate(cb, fdb, 0, 0);
+
+        char in_buf[40];
+        /* Calling ceph_lazyio_synchronize here will enable ca to fetch the propagated writes of cb in the subsequent read */
+        ceph_lazyio_synchronize(ca, fda, 0, 0);
+        ceph_read(ca, fda, in_buf, sizeof(in_buf), 0);
+        
+        /* cb does not need to call ceph_lazyio_synchronize here because it is the latest writer and the writes before it have already been propagated*/
```

Comment:
This program may be a little confusing as an example since it has a single program driving two clients. In practice, you'd almost never do this.

What may be better is to separate this into two programs that each drive their own client, maybe with a mutex or something shared between them to coordinate them?

Reply:
Ack. That sounds better. I will update it

Reply:
You should also note that these propagate/synchronize methods are not yet available for kernel or ceph-fuse mounts. LazYIO can be enabled in those environments but you cannot manage the cache by the application.

@jtlayton do we have a tracker for adding propagate/synchronize support to the kclient?

Reply:
There isn't one currently. I have some concerns about introducing this for the kclient. If you open a tracker for it, then we can discuss it there though.